test/integration/federation/framework/api.go

```diff
@@ -0,0 +1,140 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package framework
+
+import (
+	"fmt"
+	"net"
+	"net/http"
+	"strconv"
+	"testing"
+	"time"
+
+	"github.com/pborman/uuid"
+
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/kubernetes/federation/cmd/federation-apiserver/app"
+	"k8s.io/kubernetes/federation/cmd/federation-apiserver/app/options"
+	"k8s.io/kubernetes/test/integration/framework"
+)
+
+const apiNoun = "federation apiserver"
+
+func getRunOptions() *options.ServerRunOptions {
+	r := options.NewServerRunOptions()
+	r.Etcd.StorageConfig.ServerList = []string{framework.GetEtcdURLFromEnv()}
+	// Use a unique prefix to ensure isolation from other tests using the same etcd instance
+	r.Etcd.StorageConfig.Prefix = uuid.New()
+	// Disable secure serving
+	r.SecureServing.ServingOptions.BindPort = 0
+	// The insecure port will be set to an ephemeral port before starting the api
+	r.InsecureServing.BindPort = 0
+	return r
+}
+
+// FederationAPIFixture manages a federation api server
+type FederationAPIFixture struct {
+	Host     string
+	stopChan chan struct{}
+}
+
+func (f *FederationAPIFixture) Setup(t *testing.T) {
+	if f.stopChan != nil {
+		t.Fatal("Setup() already called")
+	}
+	f.stopChan = make(chan struct{})
+
+	runOptions := getRunOptions()
+
+	maxAttempts := 3
+	err := startServer(t, runOptions, f.stopChan, maxAttempts)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	f.Host = fmt.Sprintf("http://%s:%d", runOptions.InsecureServing.BindAddress, runOptions.InsecureServing.BindPort)
+
+	err = waitForServer(t, f.Host)
+	if err != nil {
+		t.Fatal(err)
+	}
+}
+
+func (f *FederationAPIFixture) Teardown(t *testing.T) {
+	close(f.stopChan)
+}
+
+func startServer(t *testing.T, runOptions *options.ServerRunOptions, stopChan <-chan struct{}, maxAttempts int) error {
+	// Try to bind to an ephemeral port, making multiple attempts if necessary.
+	// It would be preferable to use httptest.NewServer to ensure a port was
+	// bound in one step, but that will require refactoring the Run method to
+	// not require the port be known before the server is started.
+	attempts := 0
+	for attempts < maxAttempts {
+		port, err := getEphemeralPort(runOptions.InsecureServing.BindAddress)
+		if err != nil {
+			// Only count attempts to start the api
+			t.Logf("Error allocating an ephemeral port: %v", err)
+			continue
+		}
+
+		runOptions.InsecureServing.BindPort = port
+
+		err = app.NonBlockingRun(runOptions, stopChan)
```

Comment:
``Run`` blocking indefinitely makes sense for a process whose main purpose will be to run a server.  It makes it easy to block ``main`` from exiting and stopping non-``main`` goroutines.  For testing, however, async invocation adds indirection that make it more difficult to know the state of the server.  In the event of failure the server returns an error, but there is no indication of when a server has successfully bound to the specified port and can be connected to.  By running non-blocking, it's possible to tell when a client check is appropriate.  Otherwise, since there's a non-zero chance that a different process could bind to the ephemeral port between the time it is selected and when the server starts, a client check over http could end up targeting an unintended process.

Being able to run non-blocking will also make it easier to have the generic server take responsibility for ephemeral port allocation as per #42224.  When a call returns without an error, ``EffectiveSecurePort``/``EffectiveInsecurePort`` could be used to determine the ephemeral ports that were bound.